packages/js/product-editor/src/hooks/use-product-template/use-product-template.ts

```diff
@@ -38,10 +52,35 @@ export const useProductTemplate = (
 	);
 
 	if ( ! matchingProductTemplate ) {
-		// Fallback to the first template with the same product type.
-		matchingProductTemplate = productTemplates.find(
+		// Look for matching templates based on product data described on each template.
+		const matchingTemplates = productTemplates.filter(
 			( productTemplate ) =>
-				productTemplate.productData.type === productTypeToFind
+				Object.entries( productTemplate.productData ).every(
+					( [ key, value ] ) => {
+						if ( ! product ) {
+							return false;
+						}
```

Comment:
I'd move this check outside of the `every`. Without a `product`, you can't do any of the fallback matching.